lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp

```diff
@@ -418,6 +418,116 @@ class RecomposeChunkListUnpack : public OpRewritePattern<PrimListUnpackOp> {
+    Value index = op.getIndices();
+    AtenNonzeroNumpyOp nonzeroNumpyOp =
+        index.getDefiningOp<AtenNonzeroNumpyOp>();
+    if (!nonzeroNumpyOp) {
+      return rewriter.notifyMatchFailure(op,
+                                         "index of aten::index.tensor should "
+                                         "come from aten::nonzero_numpy op");
+    }
+
+    Value mask = nonzeroNumpyOp.getSelf();
+    auto inputType = input.getType().cast<BaseTensorType>();
+    auto maskType = mask.getType().cast<BaseTensorType>();
+
+    // Because this pass is run before reduce-op-variants and
+    // maximize-value-semantics pass, operation operands might already be
+    // converted through static_info_cast or copy.to_tensor op. We need to
+    // restore the shape information to judge whether we can convert the op.
+    auto originalMaskInfo = getOriginalValue(mask);
+    auto originalInputInfo = getOriginalValue(input);
+    if (failed(originalMaskInfo) || failed(originalInputInfo))
+      return rewriter.notifyMatchFailure(
+          op, "failed to get shape information of inputs of aten::index.tensor "
+              "and aten::nonzero_numpy");
+
+    auto originalMaskType =
+        (*originalMaskInfo).getType().cast<BaseTensorType>();
+    auto originalInputType =
+        (*originalInputInfo).getType().cast<BaseTensorType>();
+
+    auto originalInputSizes = originalInputType.getSizes();
+    auto originalMaskSizes = originalMaskType.getSizes();
+    if (originalMaskSizes.size() != originalInputSizes.size()) {
+      return rewriter.notifyMatchFailure(
+          op, "inputs of aten::index.tensor and aten::nonzero_numpy should "
+              "have same-ranked shape");
+    }
+
+    for (int64_t i = 0; i < originalInputSizes.size(); ++i) {
+      if (originalInputSizes[i] != originalMaskSizes[i] ||
+          originalInputSizes[i] == ShapedType::kDynamic ||
+          originalMaskSizes[i] == ShapedType::kDynamic) {
+        return rewriter.notifyMatchFailure(
+            op, "inputs of aten::index.tensor and aten::nonzero_numpy should "
+                "have same static shape");
+      }
+    }
+
+    Value falseValue = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+    Value noneValue = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value i1DtypeValue =
+        getDtypeIntValueForType(rewriter, loc, rewriter.getI1Type());
+    Type boolMaskType = maskType.getWithSizesAndDtype(
+        inputType.getOptionalSizes(), rewriter.getI1Type());
+    Value boolMask =
+        rewriter.create<AtenToDtypeOp>(loc, boolMaskType, mask, i1DtypeValue,
+                                       falseValue, falseValue, noneValue);
+
+    rewriter.replaceOpWithNewOp<AtenMaskedSelectOp>(op, op.getType(), input,
+                                                    boolMask);
+
+    // Separately handle possible prim::ListUnpack op, if it's a user of the
+    // aten::nonzero_numpy op.
+    if (nonzeroNumpyOp.getResult().hasOneUse()) {
```

Comment:
A couple of things on this:

1. Can you add a lit test for this case?
2. This should be a separate pattern, since it is fully independent of the `AtenIndexTensorOp`